bokeh/cli/core.py

```diff
@@ -164,24 +242,38 @@ def update_source(self, new_source):
         ns.index = [len_source]
         self.source = pd.concat([self.source, ns])
 
-        _c = create_chart(self.series, ns, self.index, self.factories, **self.chart_args)
-        _c._setup_show()
-        _c._prepare_show()
-        _c._show_teardown()
-
-        for k, v in _c.source.data.items():
-            self.chart.source.data[k] = list(self.chart.source.data[k]) + list(v)
-
-        chart = self.chart.chart
-        plot = chart.plot
-        plot.y_range.end = max(
-            plot.y_range.end, _c.chart.plot.y_range.end
-        )
-        plot.y_range.start = min(
-            plot.y_range.start, _c.chart.plot.y_range.start
-        )
-        chart.session.store_objects(self.chart.source)
-        chart.session.store_objects(plot)
+        # TODO: This should be replaced with something that just computes
+        #       the new data and source
+        fig = create_chart(self.series, ns, self.index, self.factories,
+                          self.map_options, **self.chart_args)
+
+        for i, _c in enumerate(fig.charts):
+            if not isinstance(_c, bc.GMap):
+                # TODO: nested charts are getting ridiculous. Need a better
+                #       better interface for charts :-)
+                scc = self.chart.charts[i]
+                for k, v in _c.source.data.items():
+                    scc.source.data[k] = list(scc.source.data[k]) + list(v)
+
+                self.limit_source(scc.source.data)
+                chart = scc.chart
+                chart.session.store_objects(scc.source)
+
+                if self.update_ranges:
+                    plot = chart.plot
+                    plot.y_range.end = max(
```

Comment:
just a minor... but why end before start? if we read the code from top to bottom, it makes sense backwards :wink:
but it is a minor as I said...